propagate configuration from paste.ini

luismsgomes opened the pull request:
Here's an example paste.ini:

```
[composite:main]
use = egg:Paste#urlmap
/pypi/ = pypiserver

[app:pypiserver]
use = egg:pypiserver#main
root = %(here)s/packages
authenticated = list update download
redirect_to_fallback = False
password_file = %(here)s/htpasswd.txt
log_file = %(here)s/log.txt
verbosity = 2

[server:main]
use = egg:gunicorn#main
host = 127.0.0.1
port = 8001
workers = 4
accesslog = %(here)s/accesslog.txt
```

The following configuration keywords were not being propagated into the app:

```
authenticated = list update download
password_file = %(here)s/htpasswd.txt
log_file = %(here)s/log.txt
verbosity = 2
```

Now it works.
Hope you like the code.
I tried to keep same style.

mplanchard commented on pypiserver/__init__.py:
```diff
@@ -139,10 +139,27 @@ def str2bool(s, default):
+    def upd_conf_with_list_item(conf, attr, sdict, sep=' ', parse=str_strip):
+        values = sdict.pop(attr, None)
+        if values:
+            conf[attr] = list(filter(None, map(parse, values.split(sep))))
```
Suggestion: use list comprehension here, as it is easier to read and negates the need to get the Python2/3-specific strip function on line 143:

```python
conf[attr] = [v.strip() for v in values.split(sep) if v]
````

luismsgomes replied:
Yes, it is easier to read but your list comprehension is not exactly equivalent to the code it replaces.
To make it equivalent, we must call parse() twice for each value:

conf[attr] = [parse(v) for v in values.split(sep) if parse(v)]

That's why I opted for the list/map/filter combination.

luismsgomes replied:
(Please let me know if you prefer the list comprehension with the two parse() calls)

mplanchard commented on pypiserver/__init__.py:
```diff
@@ -151,12 +168,23 @@ def _make_root(root):
+    upd_conf_with_str_item(c, 'log_req_frmt', local_conf)
+    upd_conf_with_str_item(c, 'log_res_frmt', local_conf)
+    upd_conf_with_str_item(c, 'log_err_frmt', local_conf)
+    upd_conf_with_str_item(c, 'welcome_file', local_conf)
```
It would tighten this up a bit and make adding new values easier to pull the item types out into a list and run the update for all of them, at least for strings, e.g.:

```python
str_items = [
    'fallback_url',
    'password_file',
    'hash_algo',
    'log_file',
    'log_frmt',
    'log_req_frmt',
    'log_res_frmt',
    'log_err_frmt',
    'welcome_file'
]

for str_item in str_items:
    upd_conf_with_str_item(c, str_item, local_conf)
````

luismsgomes replied:
Yes, that's cleaner. Thanks.

mplanchard commented on pypiserver/__init__.py:
```diff
@@ -151,12 +168,23 @@ def _make_root(root):
+    upd_conf_with_str_item(c, 'welcome_file', local_conf)
+    # cache_control is undocumented; don't know what type is expected:
+    # upd_conf_with_str_item(c, 'cache_control', local_conf)
+    print(c)
```
We should consider removing this `print()` statement to avoid people getting unexpected `stdout` output.

luismsgomes replied:
Sorry.  That print() slipped into the commit.

luismsgomes replied:
Could you please confirm if 'cache_control' should be a str? Or is it something else?

mplanchard commented:
Thanks for the contribution! I made a few comments throughout the code, and I'd also like to see a test for this before merging. If you're comfortable with `py.test`, feel free to contribute a test. If not, let me know, and I will write one.

luismsgomes commented:
Sorry, I'm not familiar with py.test yet (or other testing frameworks for Python).  That is something I want to learn, but haven't find the time, yet.  I'm leaning towards unittest for my own projects (because it is part of the standard library more than anything else).

mplanchard commented:
@luismsgomes, would you mind rebasing onto master and pushing this branch so we can verify whether the test failures are due to new things or issues that have already been resolved?

luismsgomes commented:
Hi, I will try my best.  I'm not yet familiar with git rebase (my primary versioning system is mercurial).

ankostis commented:
Dear @luismsgomes qnother option is to gradually cherry-pick your commits on top of master,
so that you can resolve any conflicts more relaxed:

```
git checkout -b  proppaste  origin/master

git cherry-pick ed0985f
git cherry-pick a4229f2
git cherry-pick 4adfd62

git branch master -f

git push <your-remote> master
```

Tip: in general, when creating PRs, it is better to create a separate branch (i.e. `proppaste`), and push that branch, instead of your master.  Otherwise, you have bound you master for this PR only, and cannot have it follow the upstream evolution.  
Note that if you decide this time to push `proppaste` instead of `master`, then you have to open a new GitHub PR, close this one, and add a comment to the new one referring to the discussion here.
I hope I did not confused you further :-)

luismsgomes commented:
Dear @ankostis, I might have messed things because I don't quite know how git rebase or cherry-pick works...  I did a pull from your master repository and git said it was replaying my changes on top of it. It didn't ask me to merge anything.  I did a push and now github says my fork is 22 commits ahead and 22 commits behind, which I don't understand at all.
Please forgive if I have messed things. I really want to learn git, but right now I have a pressing deadline to deliver my phd thesis and thus I can't take my time to learn git now.
Hope you understand.

ankostis commented:
No prob, I have fixed it for you.

Assuming you are on your `master` and have no un-committed changes in your working dir,
just execute the following commands:

```
git remote add ankostis https://luismsgomes@github.com/ankostis/pypiserver.git
git fetch ankostis
git reset --hard ankostis/proppaste
git log -1
```
And now your master should point in this commit:
```
git log -1
commit 86cbda1719c4aebc7579179542e8dd487c37da04
Author: Luís Gomes <luismsgomes@gmail.com>
Date:   Thu Jul 14 09:51:32 2016 +0100

    minor refactoring; removed extraneous print()
```

Then just ``git push <your-remote> master -f``.

ankostis commented on requirements/dev.pip:
```diff
@@ -15,7 +15,7 @@ pytest>=2.3
 mock;           python_version <= '3.2'
 gevent>=1.1b4;  python_version >= '3'
-twine>=1.6.1
+twine>=1.7,<=1.7.4
```
Why is this pinning?

mplanchard replied:
So, this shouldn't be registering as a change, and I don't think it would be with a proper rebase. This is from PR #158, which has already been merged. It was originally added because `twine` 1.8 breaks our `test_server` tests due to a new call signature.

ankostis commented:
Dear @mplanchard  I've just finished my 6-month effort to "release", so I haven't fully recovered my slices of free time for this project:-)
Can you make sense out of this PR?
Take my 86cbda17 commit if it helps you.

luismsgomes commented:
Done.
Thank you a lot @ankostis!

On Wed, Feb 15, 2017 at 4:36 PM, Kostis Anagnostopoulos <
notifications@github.com> wrote:

> No prob, I have fixed it for you.
>
> Assuming you are on your master and have no un-committed changes in your
> working dir,
> just execute the following commands:
>
> git remote add ankostis https://luismsgomes@github.com/ankostis/pypiserver.git
> git fetch ankostis
> git reset --hard ankostis/proppaste
> git log -1
>
> And now your master should point in this commit:
>
> git log -1
> commit 86cbda1719c4aebc7579179542e8dd487c37da04
> Author: Luís Gomes <luismsgomes@gmail.com>
> Date:   Thu Jul 14 09:51:32 2016 +0100
>
>     minor refactoring; removed extraneous print()
>
> Then just git push <your-remote> master -f.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypiserver/pypiserver/pull/156#issuecomment-280062595>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AItRGZ0OvkQlhuOjxGK54pJAyJOuUQZEks5rcyl2gaJpZM4JBs-S>
> .
>

mplanchard commented:
@luismsgomes, could you verify that my rebased PR #169 contains the all the changes you would expect?

I pulled in master's history by running

```
git checkout master && git pull
git checkout luismsgomes-master
git rebase -i master
```

Otherwise, it is an exact copy of your changes
